accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -1248,13 +1248,122 @@ the `LEGACY_ERASURE` of the types.
+are equal using the definition of runtime type equality specified above.
+
+The effect of this definition is to ensure that constant type literals which
+look identical in the source syntax but which may differ by the presence of
+legacy type modifiers are canonicalized consistently in the sense that any two
+type literals which would compare equal via the definition of runtime type
+equality given above will compare as identical.  For runtime implementations
+which implement identity by choosing a canonical representative for the
+equivalence class of equal instances, the choice of what type object to
+canonicalize to is arbitrary in that placement of legacy modifiers in type
+literals is not otherwise observable in the language.
+
+Note that the choice of canonicalization for type literals does not depend
+directly on whether sound or unsound null safety semantics are in use.
+
+#### Constant instances
+
+In both sound and unsound null checking, and in both opted in and opted out
+code, comparison of constant instances for identity is defined such that any two
+instances which are otherwise identical except for their generic type arguments
+shall be considered identical if those generic type arguments compare equal
+using the definition of runtime type object equality defined above.  That is,
+comparison (or canonicalization) of constant instances of generic classes is
+performed relative to the normal forms of their generic type arguments, and
+ignoring legacy type annotations as described above.  Hence, an instance of
+`C<T0>` compares identical to `C<T1>` if `T0` and `T1` have the same normal form
+(up to the identity of bound variables), and the objects are otherwise
+identical.
+
+Implementations of the Dart runtime semantics rely on canonicalization of
+constant objects to allow the identity semantics specified above to be
+implemented as fast pointer equality checks on the reference to the canonical
+form.  The definition above defines equivalence classes of constant objects for
+which we must choose the canonical representative.  The choice of this
+representative is observable in mixed mode programs, since instances with
+different degrees of "legacy-ness" in their type arguments are considered
+identical, but may contain operations which perform casts and instance checks
+which will evaluate differently depending on whether a legacy type or a
+non-legacy type is used in the canonical representative.  For example:
+
+```dart
+// null safe code.
+class C<T> {
+  void test(Object? o) {
+    assert(o is T);
+  }
+  const C(Object? o) : assert(o is T);
+}
+
+// If the canonical instance uses `int`, this is a compile time error
+// If the canonical instance uses `int*`, this is not a compile time error
+const c1 = C<int>(null);
+
+// If the canonical instance uses `int`, this throws
+// If the canonical instance uses `int*`, this does not throw
+void test1() => c1.test(null);
```

Comment:
I would declare a valid `c2` constant in this second example, since `c1` does not compile, e.g. `const c2 = C<int>(0);`